api.scm

```diff
@@ -41,12 +41,28 @@
  (oll-core internal music-tools)
  (lilypond-export lily)
  (lilypond-export MusicXML)
- (lilypond-export Humdrum)
+ ;(lilypond-export Humdrum)
```

Comment:
@PaulMorris thank you for looking into this. Now that I see it I can't believe that I seem to have managed to miss that combination. But yet, it seems it works properly now, and I'll be able to remove the "WIP" from this PR's title ASAP (was completely away over the weekend and will need some time to catch up, though).

Reply:
Glad to help @uliska .  I had another thought about this.  So far only the humdrum module is exporting an 'export-lilypond' function, but when musicxml does as well (and eventually MEI, etc.), will we end up with naming collisions where the api module doesn't know which 'export-lilypond' function to call?  I may be missing something, but it would be worth testing this out before merging.  (Which may well have been your next step!)

Reply:
No, this is by design. Until now *only* the Humdrum module has been modified, and once that works I'll make *all* modules behave that way. Actually that's the requirement for plugging in a new exporter: providing an `export-lilypond` function. The actual idea is that `load-exporter` is configured to load the appropriate module and then takes the `export-lilypond` procedure from that module.

Reply:
> See also Guile 1.8 docs on creating modules and using modules.

I had actually read this and still was confused, and I'd say the problem at hand isn't explained in a way that I would be able to really understand it.

The Guile 1.8 docs on [Loading from file](https://www.gnu.org/software/guile/docs/docs-1.8/guile-ref/Loading.html) state that `load-from-file` does "Load *filename* and evaluate its contents in the top-level environment".
What is the "top-level environment" here? Wouldn't that imply the result of a `define` or `define-public` should be visible in the file that loads the module? =>
If `api.scm'` "loads" `humdrum.scm` and `export-lilypond` is defined in `humdrum.scm` I would expect the code in `api.scm` (after the loading) to see `export-lilypond`. Is there a way to make that work without having to explicitly define a module in `humdrum.scm` and *use* it in `api.scm`?
Is `primitive-load` anything different than `load` and `load-from-file` ir is the absence of the optional `reader` argument the only difference?

My goal would be to make it possible that a new exporter is added by *only* adding a new exporter file, *without* the need to update `api.scm`.

Reply:
Here is some code:
```
\version "2.19.82"
\include "oll-core/package.ily"
% I don't load the lilypond-export package here to show the import from a list

% force loading of lilypond-export_api_module - this is not needed, if the base package is already loaded
#(display (@ (lilypond-export api) moment->duration))

% this list shall be generated with a file walk through the export-module folder
exporters.MusicXML = #'exportMusicXML
exporters.Humdrum = #'exportHumdrum

% now fetch the 
mods = #'()
#(for-each
  (lambda (p)
     ; the (cdr p) may be replaced by mandatory const like 'lilypond-export
    (let ((exporter (module-ref (resolve-module `(lilypond-export ,(car p))) (cdr p)) ))
      ; add this to the list of exporters
      (set! mods `(,@mods ,exporter)) ; use (cons key exporter)
      )) exporters)

% log found exporters with its file-suffix
#(for-each
  (lambda (exporter)
    ; when the exporter is defined with a mandatory name the name is stored in an object-property
    (ly:message "~A ~A" (procedure-name exporter) (object-property exporter 'file-suffix))
    ) mods)
```
So you can collect all files in folder, import a function `lilypond-export`  and add it to an alist (or some other kind of store). That way you can add exporters by dropping them in a directory - as long as the module follows the mandatory rules.

When my schedule allows I will prepare some code fitting into this PR.

Reply:
I only have a vague idea after a first look, but there's one thing I'd definitely *not* want to do: produce a list of all available exporter functions. These functions are already large and will become huge when they will become more comprehensive, so we should really only load *one* such function.

Therefore we may produce a list of available *module names* but the actual module loading and function parsing should occur only once.

I think what I'm missing is somehow hidden in the `(let ((exporter (module-ref (resolve-module ` line?

Reply:
If I'm not mistaken this approach doesn't help either with my problem. `resolve-module` allows loading a module by its name, but this seems to require the module being "used" too - so there's no *real* advantage over using `load-from-path`.